### PR TITLE
[prototype] return AssetMaterialization

### PR DESCRIPTION
**Pull request by alangenfeld**
explorations for returning AssetMaterializtion directly

### CODE REVIEW

**Comment by alangenfeld** [comment]
Current dependencies on/for this PR:
* master
  * **PR #14931** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/14931" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/14931?utm_source=stack-comment).

**Comment by sryza** [comment]
Have you considered a new `MaterializationMetadata` type?

**Comment by alangenfeld** [comment]
>  Have you considered a new MaterializationMetadata type?

Ya thats the other option, in previous discussions no one was hot on the idea of adding a new noun. Whats your take?

**Comment by sryza** [comment]
My initial lean would be towards `MaterializationMetadata` I think. allowing `AssetMaterialization` to be in two different states (with asset key or without) feels weird.

It's also a breaking change, right? For anyone depending on the asset key attribute to be non-None.

And it means that anyone using type-checking needs to guard against None whenever they access `AssetMaterialization.asset_key`.

**Comment by alangenfeld** [comment]
> allowing AssetMaterialization to be in two different states (with asset key or without) feels weird.

Agree, and that is not the case in the current updated state of this prototype PR. We allow omitting from the constructor in cases we can infer it and error otherwise [2].

File: python_modules/dagster/dagster/_core/definitions/events.py [hunk 1]
```diff
@@ -490,18 +491,32 @@ class AssetMaterialization(
...
+                raise DagsterInvariantViolationError(
+                    "Could not infer asset_key, not currently in the context of an execution."
+                )
+            keys = current_ctx.selected_asset_keys
+            if len(keys) != 1:
+                raise DagsterInvariantViolationError(
+                    f"Could not infer asset_key, there are {len(keys)} in the current execution"
+                    " context. Specify the appropriate asset_key."
+                )
+            asset_key = next(iter(keys))
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/definitions/events.py:519]: [2]

**Comment by alangenfeld** [comment]
How do you feel about [2] and with that does that change your stance on `MaterializationMetadata`?

**Comment by sryza** [comment]
> How do you feel about [2] and with that does that change your stance on MaterializationMetadata?

Ah, I missed this. I'm having trouble putting my finger on any concrete negatives, but this kind of pattern gives me some generalized discomfort – I think it's difficult for users to look at the code and guess how it's working.

**Comment by alangenfeld** [comment]
>  this kind of pattern gives me some generalized discomfort – I think it's difficult for users to look at the code and guess how it's working.

Makes sense, this is definitely a "may be lesser of two evils" type of proposal. 

Probably need to flesh this prototype out to include `multi_asset` as I think the full picture including how things work there will influence the final call.

File: python_modules/dagster/dagster/_core/execution/plan/execute_step.py [hunk 2]
```diff
@@ -165,6 +171,15 @@ def _step_output_error_checked_user_event_sequence(
                     f'Emitting implicit Nothing for output "{step_output_def.name}" on {op_label}'
                 )
                 yield Output(output_name=step_output_def.name, value=None)
+
+            if step_context.is_sda_step and step_context.get_observed_user_asset_mat(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/execution/plan/execute_step.py:175]: Not sure if this has any consequence with the way that execution currently works, but imagine a world where we could kick off downstream steps as soon as an upstream output is produced, even if the upstream step hasn't finished yet.

In that world, we would ideally yield the `Output` as soon the materialization is reported/yielded/returned, right?

**Pull request closed without merging**